app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java

```diff
@@ -185,38 +151,13 @@ private void shutdown() {
             Log.e("Clash", "clash stopped");
         }
         Tun2socks.stop();
+        stopSelf();
+
+        setState(STOPPED);
+        instance = null;
     }
 
     public void stop() {
         shutdown();
     }
-
-    class PacketThread extends Thread {
-        private static final String TAG = "TunPacketThread";
-
-        public void run() {
-            Log.e(TAG, Thread.currentThread().getName() + " thread start");
-            while (state == STARTING || state == STARTED) {
-                try {
-                    int n = inputStream.read(packetBuffer.array());
-                    if (n > 0) {
-                        packetBuffer.limit(n);
-                        Tun2socks.inputPacket(packetBuffer.array());
-                        packetBuffer.clear();
-                    }
-                } catch (IOException e) {
```

Comment:
The tun device could be closed by system(tun off VPN in system setting). The catch here can handle it. Is this situation considered in your design?

Reply:
https://developer.android.com/reference/android/net/VpnService#onRevoke()

The issue should be solved via commit "close resources on revoke"